contracts/LimitOrderSwap.sol

```diff
@@ -0,0 +1,207 @@
+                Asset.transferTo(Constant.ETH_ADDRESS, payable(msg.sender), ethRefund);
+            }
+        } else {
+            if (msg.value != 0) revert InvalidMsgValue();
+            _collect(order.takerToken, msg.sender, order.maker, takerSpendingAmount, takerParams.takerTokenPermit);
+        }
+
+        // avoid stack too deep error
+        _emitLimitOrderFilled(order, orderHash, takerSpendingAmount, makerSpendingAmount - fee, fee, takerParams.recipient);
+    }
+
+    function _validateOrderAndQuote(
+        LimitOrder calldata _order,
+        bytes calldata _makerSignature,
+        uint256 _takerTokenAmount,
+        uint256 _makerTokenAmount,
+        bool _fullOrKill
+    )
+        internal
+        returns (
+            bytes32 orderHash,
+            uint256 takerSpendingAmount,
+            uint256 makerSpendingAmount
+        )
+    {
+        // validate the constrain of the order
+        if (_order.expiry <= block.timestamp) revert ExpiredOrder();
+        if (_order.taker != address(0) && msg.sender != _order.taker) revert InvalidTaker();
+
+        // validate the status of the order
+        orderHash = getLimitOrderHash(_order);
+        if (orderHashToCanceled[orderHash]) revert CanceledOrder();
+
+        // check whether the order is fully filled or not
+        uint256 orderFilledAmount = orderHashToMakerTokenFilledAmount[orderHash];
+        if (orderFilledAmount >= _order.makerTokenAmount) revert FilledOrder();
+
+        // validate maker signature only once per order
+        if (orderFilledAmount == 0) {
+            if (!SignatureValidator.isValidSignature(_order.maker, getEIP712Hash(orderHash), _makerSignature)) revert InvalidSignature();
+        }
+
+        // get the quote of the fill
+        uint256 orderAvailableAmount = _order.makerTokenAmount - orderFilledAmount;
+        if (_makerTokenAmount > orderAvailableAmount) {
+            // the requested amount is larget than fillable amount
+            if (_fullOrKill) {
+                revert NotEnoughForFill();
+            } else {
+                // take the rest of this order
+                makerSpendingAmount = orderAvailableAmount;
+
+                // re-calculate the amount of taker willing to spend for this trade by the requested ratio
+                _takerTokenAmount = ((_takerTokenAmount * makerSpendingAmount) / _makerTokenAmount);
+            }
+        } else {
+            // the requested amount can be statisfied
+            makerSpendingAmount = _makerTokenAmount;
+        }
+        uint256 minTakerTokenAmount = ((makerSpendingAmount * _order.takerTokenAmount) / _order.makerTokenAmount);
+        // check if taker provide enough amount for this fill (better price is allowed)
+        if (_takerTokenAmount < minTakerTokenAmount) revert InvalidTakingAmount();
+        takerSpendingAmount = _takerTokenAmount;
+
+        if (takerSpendingAmount == 0 && makerSpendingAmount == 0) revert ZeroTokenAmount();
```

Comment:
My original thought is `AND` since it's possible for some order that one of these amount is zero. It may make no sense but would be no harm to have such flexibility. But if both are all zero than there's no need to proceed.